pymc3/variational/approximations.py

```diff
@@ -0,0 +1,432 @@
+        _log_p_D_ = tt.sum(
+            list(map(self.weighted_likelihood, self.model.observed_RVs))
+        )
+        _log_p_D_ = self.to_flat_input(_log_p_D_)
+        samples = self.sample_over_space(posterior, _log_p_D_)
+        return samples
+
+    def weighted_likelihood(self, var):
+        """Weight likelihood according to given population size
+        """
+        tot = self.population.get(
+            var, self.population.get(var.name))
+        logpt = tt.sum(var.logpt)
+        if tot is not None:
+            tot = tt.as_tensor(tot)
+            logpt *= tot
+            logpt /= var.size
+        return logpt
+
+    def KL_q_p_W(self, posterior):
+        """KL(q||p) samples over q
+        """
+        return self.log_q_W(posterior) - self.log_p_W(posterior)
+
+    def log_p_W(self, posterior):
+        """log_p_W samples over q
+        """
+        _log_p_W_ = self.model.varlogpt + tt.sum(self.model.potentials)
+        _log_p_W_ = self.to_flat_input(_log_p_W_)
+        samples = self.sample_over_space(posterior, _log_p_W_)
+        return samples
+
+    def apply_replacements(self, node, deterministic=False, include=None, exclude=None):
+        """Replace variables in graph with variational approximation. By default, replaces all variables
+
+        Parameters
+        ----------
+        node : node for replacements
+        deterministic : whether to use zeros as initial distribution
+            if True - zero initial point will produce constant latent variables
+        include : list - latent variables to be replaced
+        exclude : list - latent variables to be excluded for replacements
+
+        Returns
+        -------
+        node with replacements
+        """
+        if include is not None and exclude is not None:
+            raise ValueError('Only one parameter is supported {include|exclude}, got two')
+        if include is not None:
+            replacements = {k: v for k, v
+                            in self.flat_view.replacements.items() if k in include}
+        elif exclude is not None:
+            replacements = {k: v for k, v
+                            in self.flat_view.replacements.items() if k not in exclude}
+        else:
+            replacements = self.flat_view.replacements
+        node = theano.clone(node, replacements, strict=False)
+        posterior = self.posterior(zeros=deterministic)[0]
+        return theano.clone(node, {self.input: posterior})
+
+    def elbo(self, samples=1, pi=1):
+        """Output of this function should be used for fitting a model
+
+        Parameters
+        ----------
+        samples : Tensor - number of samples
+        pi : Tensor - weight of variational part
+
+        Returns
+        -------
+        ELBO samples
+        """
+        samples = tt.as_tensor(samples)
+        pi = tt.as_tensor(pi)
+        posterior = self.posterior(samples)
+        elbo = self.log_p_D(posterior) - pi * self.KL_q_p_W(posterior)
+        return elbo
+
```

Comment:
This way of weighting is specialized to Blundell's paper (uncertainty of weights...). But it does not support other weighting scheme like autoencoding VB with global and local parameters. Since BaseApproximation is the base class for variational approximation, it would be better to make weighting be more flexible. This is the reason that I implemented advi_minibatch() such that each of the log probability terms can be weighted individually.

If the above statement is unclear, I can make another PR to remedy this issue after this PR is merged.